File: .claude/shared/test-framework/templates/unit/javascript/basic_test_template.js
/**
 * Basic JavaScript Unit Test Template
 * 
 * This template provides a starting point for unit tests following TDD principles.
 * Supports both Jest and Mocha/Chai testing frameworks.
 */

// Jest example (most common)
describe('{ComponentName}', () => {
  let component;
  let mockDependency;
  
  // Test data
  const testData = {
    validInput: { key: 'value' },
    invalidInput: { key: null },
    edgeCase: { key: '' },
  };
  
  beforeEach(() => {
    // Set up fresh instances before each test
    mockDependency = jest.fn();
    component = new {ComponentName}({
      dependency: mockDependency
    });
  });
  
  afterEach(() => {
    // Clean up after each test
    jest.clearAllMocks();
  });
  
  describe('initialization', () => {
    test('should initialize with default values', () => {
      // Arrange & Act
      const instance = new {ComponentName}();
      
      // Assert
      expect(instance).toBeDefined();
      expect(instance.{property}).toBe({defaultValue});
    });
    
    test('should initialize with provided options', () => {
      // Arrange
      const options = {
        {property}: 'custom value'
      };
      
      // Act
      const instance = new {ComponentName}(options);
      
      // Assert
      expect(instance.{property}).toBe('custom value');
    });
  });
  
  describe('{method}', () => {
    test('should return expected result with valid input', () => {
      // Arrange
      const input = testData.validInput;
      const expectedResult = { result: 'success' };
      
      // Act
      const result = component.{method}(input);
      
      // Assert
      expect(result).toEqual(expectedResult);
      expect(mockDependency).toHaveBeenCalledWith(input);
      expect(mockDependency).toHaveBeenCalledTimes(1);
    });
    
    test('should throw error with invalid input', () => {
      // Arrange
      const input = testData.invalidInput;
      
      // Act & Assert
      expect(() => {
        component.{method}(input);
      }).toThrow('Invalid input');
    });
    
    test('should handle edge cases gracefully', () => {
      // Arrange
      const input = testData.edgeCase;
      
      // Act
      const result = component.{method}(input);
      
      // Assert
      expect(result).toBeDefined();
      expect(result.status).toBe('handled');
    });
    
    test('should handle async operations correctly', async () => {
      // Arrange
      const input = testData.validInput;
      mockDependency.mockResolvedValue({ async: 'result' });
      
      // Act
      const result = await component.{asyncMethod}(input);
      
      // Assert
      expect(result).toEqual({ async: 'result' });
      expect(mockDependency).toHaveBeenCalledWith(input);
    });
    
    test('should handle async errors correctly', async () => {
      // Arrange
      const input = testData.validInput;
      const error = new Error('Async operation failed');
      mockDependency.mockRejectedValue(error);
      
      // Act & Assert
      await expect(component.{asyncMethod}(input)).rejects.toThrow('Async operation failed');
    });
  });
  
  describe('property getters and setters', () => {
    test('should get property value correctly', () => {
      // Arrange
      component._internalProperty = 'test value';
      
      // Act
      const value = component.{property};
      
      // Assert
      expect(value).toBe('test value');
    });
    
    test('should set property value correctly', () => {
      // Arrange
      const newValue = 'new value';
      
      // Act
      component.{property} = newValue;
      
      // Assert
      expect(component._internalProperty).toBe(newValue);
    });
  });
  
  describe('event handling', () => {
    test('should emit events correctly', (done) => {
      // Arrange
      const eventData = { message: 'test event' };
      
      component.on('testEvent', (data) => {
        // Assert
        expect(data).toEqual(eventData);
        done();
      });
      
      // Act
      component.emit('testEvent', eventData);
    });
  });
  
  describe('mocking examples', () => {
    test('should work with mocked timers', () => {
      // Enable fake timers
      jest.useFakeTimers();
      
      // Arrange
      const callback = jest.fn();
      component.delayedOperation(callback, 1000);
      
      // Act - advance time
      jest.advanceTimersByTime(1000);
      
      // Assert
      expect(callback).toHaveBeenCalled();
      
      // Cleanup
      jest.useRealTimers();
    });
    
    test('should work with mocked modules', () => {
      // Mock entire module
      jest.mock('../external-module', () => ({
        externalFunction: jest.fn().mockReturnValue('mocked value')
      }));
      
      const { externalFunction } = require('../external-module');
      
      // Act
      const result = component.useExternal();
      
      // Assert
      expect(result).toBe('mocked value');
      expect(externalFunction).toHaveBeenCalled();
    });
  });
  
  describe('snapshot testing', () => {
    test('should match snapshot', () => {
      // Arrange
      const input = testData.validInput;
      
      // Act
      const result = component.{method}(input);
      
      // Assert
      expect(result).toMatchSnapshot();
    });
  });
  
  describe('performance testing', () => {
    test('should complete within performance limits', () => {
      // Arrange
      const largeInput = generateLargeTestData(1000);
      const startTime = performance.now();
      
      // Act
      component.{method}(largeInput);
      const duration = performance.now() - startTime;
      
      // Assert
      expect(duration).toBeLessThan(100); // milliseconds
    });
  });
});

// Mocha/Chai example (alternative)
const { expect } = require('chai');
const sinon = require('sinon');

describe('{ComponentName} (Mocha/Chai)', function() {
  let component;
  let mockDependency;
  
  beforeEach(function() {
    mockDependency = sinon.stub();
    component = new {ComponentName}({
      dependency: mockDependency
    });
  });
  
  afterEach(function() {
    sinon.restore();
  });
  
  it('should initialize correctly', function() {
    expect(component).to.exist;
    expect(component.{property}).to.equal({defaultValue});
  });
  
  it('should handle method call with valid input', function() {
    // Arrange
    const input = { key: 'value' };
    const expectedResult = { result: 'success' };
    
    // Act
    const result = component.{method}(input);
    
    // Assert
    expect(result).to.deep.equal(expectedResult);
    expect(mockDependency.calledOnce).to.be.true;
    expect(mockDependency.calledWith(input)).to.be.true;
  });
});

// Helper functions
function generateLargeTestData(size) {
  return Array.from({ length: size }, (_, i) => ({
    id: i,
    value: `value_${i}`
  }));
}

// Custom matchers (Jest)
expect.extend({
  toBeValidResponse(received) {
    const pass = 
      received !== null &&
      typeof received === 'object' &&
      'status' in received &&
      'data' in received;
    
    return {
      pass,
      message: () => pass
        ? `expected ${received} not to be a valid response`
        : `expected ${received} to be a valid response with status and data properties`
    };
  }
});

// Usage of custom matcher
test('should return valid response format', () => {
  const result = component.{method}();
  expect(result).toBeValidResponse();
});

File: .claude/shared/test-framework/templates/unit/python/basic_test_template.py
"""
Basic Python Unit Test Template

This template provides a starting point for unit tests following TDD principles.
"""

import unittest
from unittest.mock import Mock, patch, MagicMock
from typing import Any, Optional
import pytest  # Optional: for pytest users


class Test{ComponentName}(unittest.TestCase):
    """Test suite for {ComponentName}"""

    def setUp(self) -> None:
        """Set up test fixtures before each test method."""
        # Initialize the component under test
        self.{component} = {ComponentName}()

        # Set up any mocks or test data
        self.mock_dependency = Mock()
        self.test_data = {
            "valid_input": {"key": "value"},
            "invalid_input": {"key": None},
            "edge_case": {"key": ""},
        }

    def tearDown(self) -> None:
        """Clean up after each test method."""
        # Reset any global state or close resources
        pass

    # Test naming convention: test_{method_name}_{scenario}_{expected_outcome}

    def test_initialization_with_default_values_succeeds(self) -> None:
        """Test that component initializes correctly with default values."""
        # Arrange
        # (already done in setUp)

        # Act
        # (initialization happens in setUp)

        # Assert
        self.assertIsNotNone(self.{component})
        self.assertEqual(self.{component}.{property}, {expected_value})

    def test_{method}_with_valid_input_returns_expected_result(self) -> None:
        """Test {method} with valid input."""
        # Arrange
        input_data = self.test_data["valid_input"]
        expected_result = {"result": "success"}

        # Act
        result = self.{component}.{method}(input_data)

        # Assert
        self.assertEqual(result, expected_result)
        self.mock_dependency.assert_called_once_with(input_data)

    def test_{method}_with_invalid_input_raises_exception(self) -> None:
        """Test {method} with invalid input raises appropriate exception."""
        # Arrange
        input_data = self.test_data["invalid_input"]

        # Act & Assert
        with self.assertRaises(ValueError) as context:
            self.{component}.{method}(input_data)

        self.assertIn("Invalid input", str(context.exception))

    def test_{method}_with_edge_case_handles_gracefully(self) -> None:
        """Test {method} handles edge cases appropriately."""
        # Arrange
        input_data = self.test_data["edge_case"]

        # Act
        result = self.{component}.{method}(input_data)

        # Assert
        self.assertIsNotNone(result)
        # Add specific assertions for edge case handling

    @patch('{module}.{external_dependency}')
    def test_{method}_with_mocked_dependency_behaves_correctly(
        self, mock_external: MagicMock
    ) -> None:
        """Test {method} with mocked external dependency."""
        # Arrange
        mock_external.return_value = {"mocked": "response"}
        input_data = self.test_data["valid_input"]

        # Act
        result = self.{component}.{method}(input_data)

        # Assert
        mock_external.assert_called_once()
        self.assertEqual(result["source"], "mocked")

    def test_{property}_getter_returns_correct_value(self) -> None:
        """Test that {property} getter returns the correct value."""
        # Arrange
        expected_value = "test_value"
        self.{component}._internal_property = expected_value

        # Act
        result = self.{component}.{property}

        # Assert
        self.assertEqual(result, expected_value)

    def test_{property}_setter_updates_value_correctly(self) -> None:
        """Test that {property} setter updates the value correctly."""
        # Arrange
        new_value = "new_value"

        # Act
        self.{component}.{property} = new_value

        # Assert
        self.assertEqual(self.{component}._internal_property, new_value)

    # Performance test example (optional)
    def test_{method}_performance_within_acceptable_limits(self) -> None:
        """Test that {method} performs within acceptable time limits."""
        import time

        # Arrange
        large_input = self._generate_large_test_data(size=1000)
        max_duration = 1.0  # seconds

        # Act
        start_time = time.time()
        result = self.{component}.{method}(large_input)
        duration = time.time() - start_time

        # Assert
        self.assertLess(duration, max_duration)
        self.assertIsNotNone(result)

    # Helper methods
    def _generate_large_test_data(self, size: int) -> dict:
        """Generate large test data for performance testing."""
        return {f"key_{i}": f"value_{i}" for i in range(size)}

    def _assert_valid_state(self) -> None:
        """Assert that the component is in a valid state."""
        # Add custom state validation logic
        pass


# Pytest-style tests (alternative approach)
class TestPytest{ComponentName}:
    """Pytest-style test suite for {ComponentName}"""

    @pytest.fixture
    def component(self):
        """Fixture to create component instance."""
        return {ComponentName}()

    @pytest.fixture
    def mock_dependency(self):
        """Fixture for mocked dependency."""
        return Mock()

    def test_basic_functionality(self, component, mock_dependency):
        """Test basic functionality using pytest fixtures."""
        # Arrange
        component.dependency = mock_dependency

        # Act
        result = component.{method}("input")

        # Assert
        assert result is not None
        mock_dependency.process.assert_called_once()

    @pytest.mark.parametrize("input_data,expected", [
        ("valid", "success"),
        ("", "empty"),
        (None, "error"),
    ])
    def test_various_inputs(self, component, input_data, expected):
        """Test with various inputs using parametrize."""
        if expected == "error":
            with pytest.raises(ValueError):
                component.{method}(input_data)
        else:
            result = component.{method}(input_data)
            assert result == expected


# Custom assertions (optional)
class CustomAssertions:
    """Custom assertions for domain-specific testing."""

    def assertValidResponse(self, response: dict) -> None:
        """Assert that response has required fields."""
        self.assertIn("status", response)
        self.assertIn("data", response)
        self.assertIn("timestamp", response)

    def assertErrorResponse(self, response: dict, error_code: str) -> None:
        """Assert that response indicates an error."""
        self.assertEqual(response["status"], "error")
        self.assertEqual(response["error_code"], error_code)


# Test runner (if running directly)
if __name__ == '__main__':
    unittest.main()
